### PR TITLE
Add MGRS Extension from code in stactools-package for sentinel2

### PR DESCRIPTION
**Pull request by jsignell**
**Related Issue(s):**

- #778

**Description:**

Code from: https://github.com/stactools-packages/sentinel2/blob/main/src/stactools/sentinel2/mgrs.py
Example from: https://github.com/stac-extensions/mgrs/blob/main/examples/item.json

Added pytest-style tests, docs

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: pystac/extensions/mgrs.py
Comment (jsignell) [inline pystac/extensions/mgrs.py:1]: This is copied wholesale from stac-tools. I found a few minor issues in testing. I will call these out inline

File: pystac/extensions/mgrs.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        )
+
+    @property
+    def utm_zone(self) -> Optional[int]:
+        """Get or sets the latitude band of the datasource."""
+        return self._get_property(UTM_ZONE_PROP, int)
+
+    @utm_zone.setter
+    def utm_zone(self, v: Optional[int]) -> None:
+        self._set_property(UTM_ZONE_PROP, validated_utm_zone(v), pop_if_none=True)
```
Comment (jsignell) [inline pystac/extensions/mgrs.py:217]: This was `pop_if_none=False` in the original, but None is not allowed according to the spec.

File: pystac/extensions/mgrs.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        This extension can be applied to instances of :class:`~pystac.Item`.
+
+        Raises:
+
+            pystac.ExtensionTypeError : If an invalid object type is passed.
+        """
+        if isinstance(obj, pystac.Item):
+            cls.validate_has_extension(obj, add_if_missing)
+            return MgrsExtension(obj)
```
Comment (jsignell) [inline pystac/extensions/mgrs.py:236]: This has a `cast` around it in the original, but mypy didn't need it.

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 - `sort_links_by_id` to Catalog `get_child()` and `modify_links` to `get_stac_objects()` ([#1064](https://github.com/stac-utils/pystac/pull/1064))
 - `*ids` to Catalog and Collection `get_items()` for only including the provided ids in the iterator ([#1075](https://github.com/stac-utils/pystac/pull/1075))
 - `recursive` to Catalog and Collection `get_items()` to walk the sub-catalogs and sub-collections ([#1075](https://github.com/stac-utils/pystac/pull/1075))
+- MGRS Extension [#1088](https://github.com/stac-utils/pystac/pull/1088)
```
Comment (gadomski) [inline CHANGELOG.md:10]:
```suggestion
- MGRS Extension ([#1088](https://github.com/stac-utils/pystac/pull/1088))
```

**Review by gadomski — CHANGES_REQUESTED** [review]

**Review by gadomski — APPROVED** [review]

**Comment by philvarner** [comment]
Thanks for doing this @jsignell

**Pull request merged**